jose-jwt/jwe/JWE.cs

```diff
@@ -0,0 +1,330 @@
+                    {
+                        var compression = settings.Compression((string)compressionAlg);
+
+                        plaintext = compression.Decompress(plaintext);
+                    }
+                    return (Plaintext: plaintext, JoseHeaders: recipient.JoseHeader, Aad: parsedJwe.Aad);
+                }
+                catch (ArgumentException ex)
+                {
+                    exceptions.Add(ex);
+                }
+                catch (JoseException ex)
+                {
+                    exceptions.Add(ex);
+                }
+            }
+
+            if (exceptions.Select(e => (e.GetType(), e.Message)).Distinct().Count() == 1)
+            {
+                // throw the first
+                throw exceptions[0];
+            }
+            else
+            {
+                throw new JoseException($"No recipients able to decrypt.", new AggregateException(exceptions));
+            }
+        }
+
+        /// <summary>
+        /// Parses JWE token, extracts and unmarshal protected+unprotcted+per recipient headers as IDictionary<string, object>.
+        /// This method is NOT performing integrity checking. 
+        /// </summary>
+        /// <param name="jwe">JWE to decrypt.</param>
+        /// <param name="mode">serialization mode to use. Note only one recipient can be specified for compact and flattened json serialization.</param>
+        /// <param name="settings">optional settings to override global DefaultSettings</param>
+        /// <returns>List of Jose headers. For Compact and Flattened this will be length 1 and contain just the protected header. 
+        ///  For General Json this will be the Jose headers (merge of protected, unprotected and per-recipient).</returns>
+        /// <exception cref="IntegrityException">if AEAD operation validation failed</exception>
+        /// <exception cref="EncryptionException">if JWE can't be decrypted</exception>
+        /// <exception cref="InvalidAlgorithmException">if encryption or compression algorithm is not supported</exception>
+        public static IEnumerable<IDictionary<string, object>> UnsafeJoseHeaders(string jwe, SerializationMode mode = SerializationMode.smCompact, JwtSettings settings = null)
+        {
+            settings = GetSettings(settings);
+            
+            var parsedJwe = ParsedJwe.Parse(jwe, mode, settings);           
+
+            var protectedHeaders = settings.JsonMapper.Parse<Dictionary<string, object>>(
+                Encoding.UTF8.GetString(parsedJwe.ProtectedHeaderBytes));
+
+            var ret = new List<IDictionary<string, object>>();
+            foreach(var recipient in parsedJwe.Recipients)
+            {
+                ret.Add(MergeHeaders(protectedHeaders, parsedJwe.UnprotectedHeader, recipient.Header));
+            }
+            return ret;
+        }
+
+        private static IDictionary<string, object> MergeHeaders(params IDictionary<string, object>[] dicts)

```

Comment:
Can we either use existing `Dictionaries.Append()` or move `MergeHeades()` under `Dictionaries` too + add unit tests?

Reply:
- Using existing `Dictionaries.Append()`... I had wanted a more explicit fail-fast behaviour (as per comments in conversation below), if duplicate names attempted to be supplied. Hence not re-using existing. (Also prefered not to modify the passed arguments but return merged)
- Or move `MergeHeaders()` and add tests...  Sure I will look at doing that soon.

Reply:
I've just pushed a change to move it to Dictionaries and added unit tests.

Reply:
fair enough. Unit tests and we cool with it.